go-controller/pkg/ovn/address_set.go

```diff
@@ -0,0 +1,250 @@
+	}
+	as.uuid = uuid
+
+	if uuid != "" {
+		klog.V(5).Infof("New(%s) already exists; updating IPs", asDetail(as))
+		// ovnAddressSet already exists in the database; just update IPs
+		if err := as.setOrClear(); err != nil {
+			return nil, err
+		}
+	} else {
+		// ovnAddressSet has not been created yet. Create it.
+		args := []string{
+			"create",
+			"address_set",
+			"name=" + as.hashName,
+			"external-ids:name=" + as.name,
+		}
+		joinedIPs := as.joinIPs()
+		if len(joinedIPs) > 0 {
+			args = append(args, "addresses="+joinedIPs)
+		}
+		as.uuid, stderr, err = util.RunOVNNbctl(args...)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create address set %q, stderr: %q (%v)",
+				asDetail(as), stderr, err)
+		}
+	}
+
+	klog.V(5).Infof("New(%s) with %v", asDetail(as), ips)
+
+	return as, nil
+}
+
+func (as *ovnAddressSet) GetHashName() string {
+	return as.hashName
+}
+
+func (as *ovnAddressSet) GetName() string {
+	return as.name
+}
+
+func (as *ovnAddressSet) joinIPs() string {
+	list := make([]string, 0, len(as.ips))
+	for ipStr := range as.ips {
+		list = append(list, `"`+ipStr+`"`)
+	}
+	sort.Strings(list)
+	return strings.Join(list, " ")
+}
+
+// setOrClear updates the OVN database with the address set's addresses or
+// clears the address set if there are no addresses in the address set
+func (as *ovnAddressSet) setOrClear() error {
+	joinedIPs := as.joinIPs()
+	if len(joinedIPs) > 0 {
+		_, stderr, err := util.RunOVNNbctl("set", "address_set", as.uuid, "addresses="+joinedIPs)
+		if err != nil {
+			return fmt.Errorf("failed to set address set %q, stderr: %q (%v)",
+				asDetail(as), stderr, err)
+		}
+	} else {
+		_, stderr, err := util.RunOVNNbctl("clear", "address_set", as.uuid, "addresses")
+		if err != nil {
+			return fmt.Errorf("failed to clear address set %q, stderr: %q (%v)",
+				asDetail(as), stderr, err)
+		}
+	}
+	return nil
+}
+
+func (as *ovnAddressSet) AddIP(ip net.IP) error {
+	as.Lock()
+	defer as.Unlock()
+
+	ipStr := ip.String()
+	if _, ok := as.ips[ipStr]; ok {
+		return nil
+	}
+
+	klog.V(5).Infof("(%s) adding IP %s to address set", asDetail(as), ipStr)
+
+	_, stderr, err := util.RunOVNNbctl("add", "address_set", as.uuid, "addresses", `"`+ipStr+`"`)
+	if err != nil {
+		return fmt.Errorf("failed to add IP %q to address set %q, stderr: %q (%v)",
+			ip, asDetail(as), stderr, err)
+	}
+
+	as.ips[ip.String()] = ip
+	return nil
+}
+
+func (as *ovnAddressSet) DeleteIP(ip net.IP) error {
+	as.Lock()
+	defer as.Unlock()
+
+	ipStr := ip.String()
+	if _, ok := as.ips[ipStr]; !ok {
+		return nil
+	}
+
+	klog.V(5).Infof("(%s) deleting IP %s from address set", asDetail(as), ipStr)
+
+	_, stderr, err := util.RunOVNNbctl("remove", "address_set", as.uuid, "addresses", `"`+ipStr+`"`)
+	if err != nil {
+		return fmt.Errorf("failed to remove IP %q from address set %q, stderr: %q (%v)",
+			ip, asDetail(as), stderr, err)
+	}
+
+	delete(as.ips, ipStr)
+	return nil
+}
+
+func (as *ovnAddressSet) Destroy() {
```

Comment:
`Destroy()` is almost always called in conjunction with removing the address set reference for it to be garbage collected, but to ensure the integrity of the object and make it consistent with the nbdb, does it make sense to set the `as.ips = nil` after the `nbctl` call succeeds?

Reply:
done and fixed up to return errors back to callers (which then log them)